src/vmm/src/vstate/vm.rs

```diff
@@ -0,0 +1,452 @@
+    VmCreateGIC(arch::aarch64::gic::Error),
+    /// Cannot open the VM file descriptor.
+    VmFd(kvm_ioctls::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Failed to get KVM vm pit state.
+    VmGetPit2(kvm_ioctls::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Failed to get KVM vm clock.
+    VmGetClock(kvm_ioctls::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Failed to get KVM vm irqchip.
+    VmGetIrqChip(kvm_ioctls::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Failed to set KVM vm pit state.
+    VmSetPit2(kvm_ioctls::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Failed to set KVM vm clock.
+    VmSetClock(kvm_ioctls::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Failed to set KVM vm irqchip.
+    VmSetIrqChip(kvm_ioctls::Error),
+    /// Cannot configure the microvm.
+    VmSetup(kvm_ioctls::Error),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+
+        match self {
+            KvmApiVersion(v) => write!(
+                f,
+                "The host kernel reports an invalid KVM API version: {}",
+                v
+            ),
+            KvmCap(cap) => write!(f, "Missing KVM capabilities: {:?}", cap),
+            #[cfg(target_arch = "x86_64")]
+            GuestMSRs(e) => write!(f, "Retrieving supported guest MSRs fails: {:?}", e),
+            #[cfg(target_arch = "aarch64")]
+            VmCreateGIC(e) => write!(f, "Error creating the global interrupt controller: {:?}", e),
+            VmFd(e) => write!(f, "Cannot open the VM file descriptor: {}", e),
+            VmSetup(e) => write!(f, "Cannot configure the microvm: {}", e),
+            NotEnoughMemorySlots => write!(
+                f,
+                "The number of configured slots is bigger than the maximum reported by KVM"
+            ),
+            SetUserMemoryRegion(e) => write!(f, "Cannot set the memory regions: {}", e),
+            #[cfg(target_arch = "x86_64")]
+            VmGetPit2(e) => write!(f, "Failed to get KVM vm pit state: {}", e),
+            #[cfg(target_arch = "x86_64")]
+            VmGetClock(e) => write!(f, "Failed to get KVM vm clock: {}", e),
+            #[cfg(target_arch = "x86_64")]
+            VmGetIrqChip(e) => write!(f, "Failed to get KVM vm irqchip: {}", e),
+            #[cfg(target_arch = "x86_64")]
+            VmSetPit2(e) => write!(f, "Failed to set KVM vm pit state: {}", e),
+            #[cfg(target_arch = "x86_64")]
+            VmSetClock(e) => write!(f, "Failed to set KVM vm clock: {}", e),
+            #[cfg(target_arch = "x86_64")]
+            VmSetIrqChip(e) => write!(f, "Failed to set KVM vm irqchip: {}", e),
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Describes a KVM context that gets attached to the microVM.
+/// It gives access to the functionality of the KVM wrapper as
+/// long as every required KVM capability is present on the host.
+pub struct KvmContext {
```

Comment:
This needs to be defined in another file probably. It is not related to the vm, but to the system wide state. In rust-vmm/kvm-ioctls, this functionality is part of the `system` module.

Reply:
@andreeaflorescu  I applied your suggestion. Please re-review.